eman2/protocols/protocol_tomo_subtomogram_refinement.py

```diff
@@ -0,0 +1,243 @@
+        writeSetOfSubTomograms(self.inputSetOfSubTomogram.get(), storePath)
+
+    def refinementSubtomogram(self):
+
+        """ Run the Subtomogram refinement. """
+        args = ' %s' % self.newFn
+        if not isinstance(self.inputRef.get(), NoneType):
+            args += (' --reference=%s ' % self.inputRef.get().getFileName())
+        args += (' --mass=%f' % self.mass)
+        args += ' --threads=%d' % self.threads
+        args += ' --goldstandard=%d ' % self.goldstandard
+        args += ' --pkeep=%f ' % self.pkeep
+        args += ' --sym=%s ' % self.sym
+        args += ' --maxtilt=%s ' % self.maxtilt
+        args += ' --path=%s ' % self.getOutputPath()
+        if self.niter > 1:
+            args += ' --niter=%d' % self.niter
+        if self.goldcontinue:
+            args += ' --goldcontinue '
+        if self.localfilter:
+            args += ' --localfilter '
+
+        print("command: e2spt_refine.py " + args)
+        program = eman2.Plugin.getProgram('e2spt_refine.py')
+        self.runJob(program, args)
+
+    def runMLStep(self, params):
+        pass
+
+    def getLastFromOutputPath(self, pattern):
+        threedPaths = glob(self.getOutputPath("*"))
+        imagePaths = sorted(path for path in threedPaths if re.match(pattern, os.path.basename(path)))
+        if not imagePaths:
+            raise Exception("No file in output directory matches pattern: %s" % pattern)
+        else:
+            return imagePaths[-1]
+
+    def createOutputStep(self):
+        lastImage = self.getLastFromOutputPath("threed_\d+.hdf")
+        samplingRate = self.inputRef.get().getSamplingRate()
+        inputSetOfSubTomograms = self.inputSetOfSubTomogram.get()
+
+        # Output 1: Subtomogram
+        subTomogram = SubTomogram()
+        subTomogram.setFileName(lastImage)
+        subTomogram.copyInfo(inputSetOfSubTomograms)
+        subTomogram.setSamplingRate(samplingRate)
+
+        # Output 2: setOfSubTomograms
+        particleParams = getLastParticlesParams(self.getOutputPath())
+        outputSetOfSubTomograms = self._createSet(SetOfSubTomograms, 'subtomograms%s.sqlite', "")
+        outputSetOfSubTomograms.copyInfo(inputSetOfSubTomograms)
+        outputSetOfSubTomograms.setCoordinates3D(inputSetOfSubTomograms.getCoordinates3D())
+        outputSetOfSubTomograms.setSamplingRate(samplingRate) # diff
+        updateSetOfSubTomograms(inputSetOfSubTomograms, outputSetOfSubTomograms, particleParams)
+
+        self._defineOutputs(outputSubTomogram=subTomogram, outputSetOfSubTomograms=outputSetOfSubTomograms)
+        self._defineSourceRelation(self.inputSetOfSubTomogram, subTomogram)
+        self._defineSourceRelation(self.inputSetOfSubTomogram, outputSetOfSubTomograms)
+
+    def getOutputPath(self, *args):
+        return self._getExtraPath(self.OUTPUT_DIR, *args)
+
+    def getOutputFile(self, folderpattern, folder, files, pattern):
+        pattern = "^" + folderpattern + pattern
+        import re
```

Comment:
I think all the imports should go together at the begining

Reply:
They were already at the top but forgot to delete it from here as well. Done now https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487